### PR TITLE
[SDK/CLI] Support pfazure run create with remote flow

### PR DESCRIPTION
**Pull request by 0mza987**
- Support pfazure run create with remote flow.
  - For remote workspace flow: `pfazure run create --flow azureml:<flow-name>`
  - For remote registry flow: `pfazure run create --flow azureml://registries/<registry-name>/models/<flow-name>/versions/<flow-version>`

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 1]
```diff
@@ -132,9 +139,14 @@ def __init__(
         self._creation_context = kwargs.get("creation_context", None)
         # init here to make sure those fields initialized in all branches.
         self.flow = flow
+        self._use_remote_flow = is_remote_uri(flow)
         self._experiment_name = None
         self._lineage_id = None
-        if self._run_source == RunInfoSources.LOCAL:
+        if self._use_remote_flow:
+            self._flow_name = get_flow_name_from_remote_flow_pattern(flow)
+            self._experiment_name = self._flow_name
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_run.py:147]: what's the _flow_name here when _use_remote_flow?

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 2]
```diff
@@ -315,7 +327,7 @@ def _to_dict(self):
         }
 
         if self._run_source == RunInfoSources.LOCAL:
-            result["flow_name"] = Path(str(self.flow)).resolve().name
+            result["flow_name"] = Path(str(self.flow)).resolve().name if not self._use_remote_flow else self._flow_name
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_run.py:330]: maybe make this a helper function? see it many times in the file

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 3]
```diff
@@ -531,3 +546,27 @@ def _validate_and_return_run_name(run: Union[str, "Run"]) -> str:
...
+        """Validate run object for create operation."""
+        # check flow value
+        if Path(self.flow).is_dir():
+            # local flow
+            pass
+        elif isinstance(self.flow, str) and self.flow.startswith(REMOTE_URI_PREFIX):
+            # remote flow
+            pass
+        else:
+            raise ValueError(
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_run.py:560]: Use PromptflowException instead.

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 4]
```diff
@@ -531,3 +546,27 @@ def _validate_and_return_run_name(run: Union[str, "Run"]) -> str:
...
+            )
+
+        if is_remote_uri(self.data):
+            # Pass through ARM id or remote url, the error will happen in runtime if format is not correct currently.
+            pass
+        else:
+            if self.data and not Path(self.data).exists():
+                raise FileNotFoundError(f"data path {self.data} does not exist")
+        if not self.run and not self.data:
+            raise ValueError("at least one of data or run must be provided")
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_run.py:572]: Use PromptflowException instead.

File: src/promptflow/promptflow/_sdk/entities/_run.py (hunk 1 above)
Comment (0mza987) [reply src/promptflow/promptflow/_sdk/entities/_run.py:147]: It's the flow id aka flow name.

File: src/promptflow/promptflow/_sdk/entities/_run.py (hunk 3 above)
Comment (0mza987) [reply src/promptflow/promptflow/_sdk/entities/_run.py]: Updated

File: src/promptflow/promptflow/_sdk/entities/_run.py (hunk 2 above)
Comment (0mza987) [reply src/promptflow/promptflow/_sdk/entities/_run.py]: Refined the logic, use `self._flow_name` instead.

**Review by wangchao1230 — APPROVED** [review]

**Review by zhengfeiwang — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 5]
```diff
@@ -975,9 +978,37 @@ def is_url(value: Union[PathLike, str]) -> bool:
...
+        if not match or len(match.groups()) != 2:
+            raise UserErrorException(error_message)
+        flow_name, _ = match.groups()
+    # check for workspace flow pattern
+    elif flow.startswith(REMOTE_URI_PREFIX):
+        pattern = r"azureml:(?P<name>.*?)$"
+        match = re.match(pattern, flow)
+        if not match or len(match.groups()) != 1:
+            raise UserErrorException(error_message)
+        flow_name = match.groups()[0]
```
Comment (crazygao) [inline src/promptflow/promptflow/_sdk/_utils.py:1013]: if you use group name, please use the name.
or else you don't need to use the group name.

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 6]
```diff
@@ -975,9 +978,37 @@ def is_url(value: Union[PathLike, str]) -> bool:
...
+    # check for registry flow pattern
+    if flow.startswith(REGISTRY_URI_PREFIX):
+        pattern = r"azureml://registries/.*?/models/(?P<name>.*?)/versions/(?P<version>.*?)$"
+        match = re.match(pattern, flow)
+        if not match or len(match.groups()) != 2:
+            raise UserErrorException(error_message)
+        flow_name, _ = match.groups()
+    # check for workspace flow pattern
+    elif flow.startswith(REMOTE_URI_PREFIX):
+        pattern = r"azureml:(?P<name>.*?)$"
```
Comment (crazygao) [inline src/promptflow/promptflow/_sdk/_utils.py:1009]: By the way this is a little bit wired.

.*? mean you don't want greedy mode.
But $ means to match to the very end.

.* will do the work perfectly

**Review by crazygao — APPROVED** [review]

**Pull request merged**